### PR TITLE
Fork tests

### LINKED ISSUE #144: Mento fork tests
**Description**

We want to have an integration test suite that can be run against a fork of any of our environments (Baklava, Alfajores, Mainnet).
The test suites should use foundry forking and be an extension of the [`BrokerIntegration`](https://github.com/mento-protocol/mento-core/blob/develop/test/integration/BrokerIntegration.t.sol) test suite, but configured in such a way that it can run on top of any test environment

**Acceptance criteria**

- [ ] There should be a command that runs the test suite on top of a given environment
- [ ] The tests should cover the following:
  - Swaps on all of the pools can happen in both directions and the price is in +/-5% of the oracle price.
  - Trading limits are configured for each pool
  - Trading limits are correctly enforced on all pools in both directions
  - Trading limits reset after the correct amount of time and trading resumes
  - Circuit breaker is configured for each rate
  - Circuit breaker trips according to the configured thresholds for all rates
  - Circuit breaker, once tripped, resets after the cooldown time
  - Circuit breaker tripping causes trading to halt

### PR DESCRIPTION
**Pull request by bowd**
### Description

Introduce a class of tests that are designed to run on top of a fork of the mento protocol.
The tests are designed under the assumption that they can run on top of the current state of a live system, so it doesn't make any assumptions or does any setup. It verifies that the invariants of the system hold.

This PR tackles all the test cases in the Issue #144
- [x] Swaps on all of the pools can happen in both directions and the price is in +/-5% of the oracle price.
- [x] Trading limits are configured for each pool
- [x] Trading limits are correctly enforced on all pools in both directions
- [x] Trading limits reset after the correct amount of time and trading resumes
- [x] Circuit breaker is configured for each rate
- [x] Circuit breaker trips according to the configured thresholds for all rates
- [x] Circuit breaker, once tripped, resets after the cooldown time
- [x] Circuit breaker tripping causes trading to halt

> Tho' the intent is to not configure the system, the PR currently sets trading limits and circuit breaker because they weren't added to the first Baklava deployment, so they are temporarily added here in order to test the test the test.

### Other changes

- Modified remappings.txt to make the Solidity vscode extension happy
- Added the Chain.sol from mento-deployment to this repo as well for the fork tests (in the future we might want to have a mento-libs repo for shared logic but for now it's fine to duplicate these as they are test/deployment utils)
- Modified the TokenHelper to have a generic mint method. 

### Tested

Tested on top of Baklava but with manually configured trading limits and circuit breaker

### Related issues

- Fixes  #144 

### Backwards compatibility

All good.

### Documentation

All good.

### CODE REVIEW

File: test/fork-tests/TestAsserts.t.sol [hunk 1]
```diff
@@ -0,0 +1,583 @@
...
+  using SafeMath for uint256;
+  using TradingLimits for TradingLimits.State;
+  using TradingLimits for TradingLimits.Config;
+  using FixidityLib for FixidityLib.Fraction;
+
+  uint8 private constant L0 = 1; // 0b001 Limit0
+  uint8 private constant L1 = 2; // 0b010 Limit1
+  uint8 private constant LG = 4; // 0b100 LimitGlobal
+
+  bool __swapAssertDebug = false;
```
Comment (bayological) [inline test/fork-tests/TestAsserts.t.sol:36]: What is the purpose of this flag?

**Review by bayological — APPROVED** [review]
This looks good to me. Nice work 💪.

One thing that would be nice to have, is commands that allow us to run tests for each environment. Currently I need to pass in the profile as an env var when running `forge test --match-contract Baklava`. It would be nice to run a command such as `yarn fork-test:baklava` instead. WDYT?

File: test/fork-tests/TestAsserts.t.sol (hunk 1 above)
Comment (bowd) [reply test/fork-tests/TestAsserts.t.sol:36]: Will remove it. At some point, I wanted to limit the console.log depending on this flag, but in the end, I realized it didn't really make sense to do that and simply rely on the `verbose` flags in the command to control that.

**Comment by bowd** [comment]
> This looks good to me. Nice work 💪.
> 
> One thing that would be nice to have, is commands that allow us to run tests for each environment. Currently I need to pass in the profile as an env var when running `forge test --match-contract Baklava`. It would be nice to run a command such as `yarn fork-test:baklava` instead. WDYT?

I like 👍 added those in package.json.

File: test/fork-tests/TestAsserts.t.sol [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+
+  function assert_swapInFails(
+    Utils.Context memory ctx,
+    address from,
+    address to,
+    uint256 sellAmount,
+    string memory revertReason
+  ) internal {
+    ctx.t.mint(from, ctx.t.trader0(), sellAmount);
+    IERC20Metadata(from).approve(address(ctx.broker), sellAmount);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:81]: `ctx.broker.swapIn` already mints + approves so no need to do it here?

File: test/fork-tests/TestAsserts.t.sol [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+  function assert_swapOutFails(
+    Utils.Context memory ctx,
+    address from,
+    address to,
+    uint256 buyAmount,
+    string memory revertReason
+  ) internal {
+    uint256 maxAmountIn = ctx.broker.getAmountIn(ctx.exchangeProvider, ctx.exchangeId, from, to, buyAmount);
+    ctx.t.mint(from, ctx.t.trader0(), maxAmountIn);
+    IERC20Metadata(from).approve(address(ctx.broker), maxAmountIn);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:96]: ditto `ctx.broker.swapOut`

File: test/fork-tests/TestAsserts.t.sol [hunk 4]
```diff
@@ -0,0 +1,581 @@
...
+  function assert_swapOverLimitFails(
+    Utils.Context memory ctx,
+    address from,
+    address to,
+    uint8 limit
+  ) internal {
+    TradingLimits.Config memory limitConfigFrom = ctx.tradingLimitsConfig(from);
+    TradingLimits.Config memory limitConfigTo = ctx.tradingLimitsConfig(to);
+
+    // Always only one limit on a pair
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:112]: do we want to assert that this is true in `test_tradingLimitsAreConfigured`?

File: test/fork-tests/TestAsserts.t.sol [hunk 5]
```diff
@@ -0,0 +1,581 @@
...
+     * during inflow on `from`, therfore we check the positive end
+     * of the limit because `from` flows into the reserve.
+     */
+    TradingLimits.Config memory limitConfig = ctx.tradingLimitsConfig(from);
+    TradingLimits.State memory limitState = ctx.refreshedTradingLimitsState(from);
+
+    console.log(block.timestamp, "Swap until limit on from (L0)");
+    int48 maxPossible = limitConfig.limit0 - limitState.netflow0;
+    console.log("Max possible: ");
+    console.logInt(maxPossible);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:204]: nit: merge to one line?

File: test/fork-tests/TestAsserts.t.sol [hunk 6]
```diff
@@ -0,0 +1,581 @@
...
+     * during outflow on `to`, therfore we check the negative end
+     * of the limit because `to` flows out of the reserve.
+     */
+    TradingLimits.Config memory limitConfig = ctx.tradingLimitsConfig(to);
+    TradingLimits.State memory limitState = ctx.tradingLimitsState(to).update(limitConfig, 0, 0);
+
+    console.log(block.timestamp, "Swap until limit on to (L0)");
+    int48 maxPossible = limitConfig.limit0 + limitState.netflow0 - 1;
+    console.log("Max possible: ");
+    console.logInt(maxPossible);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:288]: ditto: merge to one line?

File: test/fork-tests/TestAsserts.t.sol [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+    address to
+  ) public {
+    /*
+     * from -> L0[to]
+     * This function will do valid swaps until just before L0 is hit
+     * during outflow on `to`, therfore we check the negative end
+     * of the limit because `to` flows out of the reserve.
+     */
+    TradingLimits.Config memory limitConfig = ctx.tradingLimitsConfig(to);
+    TradingLimits.State memory limitState = ctx.tradingLimitsState(to).update(limitConfig, 0, 0);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:283]: use `ctx.refreshedTradingLimitsState(to)`?

File: test/fork-tests/TestAsserts.t.sol [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+    address to
+  ) public {
+    /*
+     * from -> L1[to]
+     * This function will do valid swaps until just before L1 is hit
+     * during outflow on `to`, therfore we check the negative end
+     * of the limit because `to` flows out of the reserve.
+     */
+    TradingLimits.Config memory limitConfig = ctx.tradingLimitsConfig(to);
+    TradingLimits.State memory limitState = ctx.tradingLimitsState(to).update(limitConfig, 0, 0);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:306]: ditto use `ctx.refreshedTradingLimitsState(to)`?

File: test/fork-tests/TestAsserts.t.sol [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+
+    console.log(block.timestamp, "Swap until limit on to (L1)");
+    int48 maxPerSwap = limitConfig.limit0 - 1;
+
+    while (limitState.netflow1 - maxPerSwap > -1 * limitConfig.limit1) {
+      skip(limitConfig.timestep0 + 1);
+      swapUntilL0_onOutflow(ctx, from, to);
+      limitConfig = ctx.tradingLimitsConfig(to);
+      limitState = ctx.tradingLimitsState(to).update(limitConfig, 0, 0);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:315]: do `limitState = ctx.tradingLimitsState(to)`?

File: test/fork-tests/TestAsserts.t.sol [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+    console.log(block.timestamp, "Swap until limit on to (LG)");
+
+    if (limitConfig.isLimitEnabled(L1)) {
+      int48 maxPerSwap = limitConfig.limit0 - 1;
+      while (limitState.netflowGlobal - maxPerSwap > -1 * limitConfig.limitGlobal) {
+        skip(limitConfig.timestep1 + 1);
+        swapUntilL1_onOutflow(ctx, from, to);
+        limitConfig = ctx.tradingLimitsConfig(to);
+        // Triger an update to reset netflows
+        limitState = ctx.tradingLimitsState(to).update(limitConfig, 0, 0);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:343]: do `limitState = ctx.tradingLimitsState(to)`?

File: test/fork-tests/TestAsserts.t.sol [hunk 11]
```diff
@@ -0,0 +1,581 @@
...
+      }
+      skip(limitConfig.timestep1 + 1);
+    } else if (limitConfig.isLimitEnabled(L0)) {
+      int48 maxPerSwap = limitConfig.limit0 - 1;
+      while (limitState.netflowGlobal - maxPerSwap > -1 * limitConfig.limitGlobal) {
+        skip(limitConfig.timestep0 + 1);
+        swapUntilL0_onOutflow(ctx, from, to);
+        limitConfig = ctx.tradingLimitsConfig(to);
+        // Triger an update to reset netflows
+        limitState = ctx.tradingLimitsState(to).update(limitConfig, 0, 0);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:353]: do `limitState = ctx.tradingLimitsState(to)`?

File: test/fork-tests/TestAsserts.t.sol [hunk 12]
```diff
@@ -0,0 +1,581 @@
...
+    address to
+  ) public {
+    /*
+     * from -> LG[to]
+     * This function will do valid swaps until just before LG is hit
+     * during outflow on `to`, therfore we check the negative end
+     * of the limit because `to` flows out of the reserve.
+     */
+    TradingLimits.Config memory limitConfig = ctx.tradingLimitsConfig(to);
+    TradingLimits.State memory limitState = ctx.tradingLimitsState(to).update(limitConfig, 0, 0);
```
Comment (nvtaveras) [inline test/fork-tests/TestAsserts.t.sol:332]: ditto use `ctx.refreshedTradingLimitsState(to)`?

**Review by nvtaveras — CHANGES_REQUESTED** [review]

File: test/fork-tests/TestAsserts.t.sol (hunk 5 above)
Comment (nvtaveras) [reply test/fork-tests/TestAsserts.t.sol:204]: added revert to check is not negative

File: test/fork-tests/TestAsserts.t.sol (hunk 6 above)
Comment (nvtaveras) [reply test/fork-tests/TestAsserts.t.sol:288]: added revert to check is not negative

**Comment by nvtaveras** [comment]
LGTM 🚀

**Review by nvtaveras — APPROVED** [review]

**Pull request merged**